teach/concepts/data-validation/versions.md

```diff
@@ -0,0 +1,136 @@
+        - `dependentProperties: PropertyDescriptor[]`: An array of property descriptor objects that this rule depends on. This wiil be used for the `ValidationService` to know when to re-run the rule.
+    - **Methods**:
+        - `validate(dataInstance): Promise<ValidationError | null>`:
+            - Core asynchronous method all subclasses must implement.
+            - Accepts the data object instance for validation.
+            - Run rules in parallel using `Promise.all`.
+            - Returns a **Promise** resolving to either:
+                - A `ValidationError` object (validation fails)
+                - `null` (validation succeeds)
+    - **Pseudo-code for the base class**
+
+        ```js
+        class ValidationRule {
+            constructor(name, message, dependentProperties) {
+                this.name = name;
+                this.message = message;
+                this.dependentProperties = dependentProperties;
+            }
+
+            async validate(dataInstance) {
+                // This method must be implemented by subclasses.
+                throw new Error("ValidationRule.validate() must be implemented.");
+            }
+        }
+        ```
+
+2. `ExpressionValidationRule` Subclass (**extends `ValidationRule`)**
+
+    - **Purpose**: To validate a data instance against a given string expression.
+    - **Properties**:
+        - Inherits `name`, `message`, and `dependentProperties` from `ValidationRule`.
+        - `criteria: Criteria`: Object that contains an expression string to be evaluated (e.g., "endDate > startDate`"`).
+    - **Methods**:
+        - `validate(dataInstance): Promise<ValidationError | null>`:
+            1. **Override** the base class method.
+            2. Evaluate the `criteria` expression against the instance object.
+            3. Return a `ValidationError` if the expression is false, using the rule's details.
+            4. Return `null` if the expression evaluates to true.
+    - **Pseudo-code for the subclass**
+
+        ```js
+        class ExpressionValidationRule extends ValidationRule {
+            constructor(name, message, dependentProperties, criteria) {
+                super(name, message, dependentProperties);
+                this.criteria = criteria;
+            }
+
+            async validate(dataInstance) {
+                // delegate logic to FRB...
+                const isValid = evaluateExpression(this.criteria, dataInstance);
+
+                if (!isValid) {
+                    // Rule failed, return the error object.
+                    return new ValidationError(
+                      message: this.message, // Later, we should use a localization service.
+                      dependentProperties: this.dependentProperties,
+                      rule: this,
+                    );
+                }
+
+                // Rule passed.
+                return null;
+           }
+        }
+        ```
+
+3. `ValidationService`
```

Comment:
No, we don't need a validationService for that. It is a feature of a DataService, not a new kind of DataService.

We've been focused on Data-level validation, I don't think there's a case for raw-data level one  as raw data is the result of a RawDataService doing its business where it has full control

Reply:
I see, well, we could certainly group all validation-related methods in a separate file called data-service-validation.js, or create a DataValidationManager class to take care of that, sure, it's just "Service" shouldn't be reused for something else, that would create confusion.